File: src/main/java/com/sothawo/akkabatch/Writer.java
package com.sothawo.akkabatch;

import akka.actor.ActorSelection;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.sothawo.akkabatch.messages.*;

import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.text.MessageFormat;
import java.util.TreeMap;

/**
 * Writer Actor.
 *
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
public class Writer extends AkkaBatchActor {
// ------------------------------ FIELDS ------------------------------

    /**
     * Logger
     */
    final LoggingAdapter log = Logging.getLogger(getContext().system(), this);
    /**
     * to write the output
     */
    private PrintWriter writer;
    /**
     * Reader-actor
     */
    private ActorSelection reader;
    /**
     * Buffer als TreeMap, because we need sorted access to the data
     */
    private final TreeMap<Long, ProcessRecord> outputBuffer = new TreeMap<>();
    /**
     * recordId of the next record in the output file
     */
    private long nextRecordId;

// ------------------------ CANONICAL METHODS ------------------------

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof ProcessRecord) {
            processRecord((ProcessRecord) message);
        } else if (message instanceof InitWriter) {
            initWriter((InitWriter) message);
        } else {
            unhandled(message);
        }
    }

    /**
     * processes the next record
     *
     * @param processRecord
     *         record to write
     */
    private void processRecord(ProcessRecord processRecord) {
        Long recordId = processRecord.getRecordId();
        // if the id has already been processed, ignore it, has been resend by the Reader too often.
        if (recordId >= nextRecordId) {
            reader.tell(new RecordReceived(recordId), getSelf());
            outputBuffer.put(recordId, processRecord);

            long recordsWritten = 0;
            while (!outputBuffer.isEmpty() && nextRecordId == outputBuffer.firstKey()) {
                ProcessRecord record = outputBuffer.remove(nextRecordId);
                writer.println(record.getCsvOriginal());
                recordsWritten++;
                nextRecordId++;
                if (0 == (nextRecordId % 10000)) {
                    log.debug(MessageFormat.format("processed: {0}", nextRecordId));
                }
            }

            if (0 < recordsWritten) {
                reader.tell(new RecordsWritten(recordsWritten), getSelf());
            }
        }
    }

    /**
     * Initializes the writer
     *
     * @param message
     *         init message
     */
    private void initWriter(InitWriter message) {
        Boolean result = true;
        try {
            writer = new PrintWriter(message.getOutputFilename(), message.getEncoding());
        } catch (FileNotFoundException | UnsupportedEncodingException e) {
            log.error(e, "Initializing Writer");
            result = false;
        }
        outputBuffer.clear();
        nextRecordId = 1;
        log.info(MessageFormat.format("file: {0}, encoding: {1}, Init-result: {2}", message.getOutputFilename(),
                                      message.getEncoding(), result));
        sender().tell(new InitResult(result), getSelf());
    }

    @Override
    public void postStop() throws Exception {
        if (null != writer) {
            writer.flush();
            writer.close();
            writer = null;
        }
        super.postStop();
    }

    @Override
    public void preStart() throws Exception {
        super.preStart();
        reader = getContext().actorSelection(configApp.getString("names.readerRef"));
        log.debug(MessageFormat.format("sending infos to {0}", reader.pathString()));
    }
}

File: src/main/scala/com/sothawo/akkabatch/scala/Writer.scala
package com.sothawo.akkabatch.scala

import java.io.PrintWriter


import akka.actor.{ActorSelection, Props}
import com.sothawo.akkabatch.scala.messages._
import scala.collection.mutable
import scala.collection.immutable.TreeMap

/**
 * Writer Actor.
 *
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
class Writer extends AkkaBatchActor {
  /** to write the output */
  private var writer: PrintWriter = null

  /** Reader-actor */
  private var reader: ActorSelection = null

  /** mapping from record id to actual data as a tree map to have the recordIds ordering */
  private var outputData = TreeMap[Long, ProcessRecord]()

  /** recordId of the next record in the output file */
  private var nextRecordId: Long = 0L

  override def preStart() {
    super.preStart()
    reader = context.actorSelection(appConfig.getString("names.readerRef"))
    log.debug(s"sending infos to ${reader.pathString}")
  }

  override def postStop() {
    if (null != writer) {
      writer.flush()
      writer.close()
      writer = null
    }
    super.postStop()
  }

  def receive = {
    case msg: InitWriter => initWriter(msg)
    case msg: ProcessRecord => processRecord(msg)
  }

  /**
   * initializes the writer
   * @param msg init message
   */
  private def initWriter(msg: InitWriter) = {
    var result = true

    try {
      writer = new PrintWriter(msg.outputFilename, msg.encoding)
    }
    catch {
      case e: Exception => {
        log.error(e, "Initializing Writer")
        result = false
      }
    }

    outputData = TreeMap[Long, ProcessRecord]()
    nextRecordId = 1
    log.info(s"file: ${msg.outputFilename}, encoding: ${msg.encoding}, Init-result: ${result}")

    sender ! InitResult(result)
  }

  private def processRecord(msg: ProcessRecord) = {
    val recordId = msg.recordId
    // if the id has already been processed, ignore it, has been resend by the Reader too often.
    if (recordId >= nextRecordId) {
      reader ! RecordReceived(recordId)
      outputData += (recordId -> msg)
      var recordsWritten = 0L
      while (!outputData.isEmpty && nextRecordId == outputData.firstKey) {
        val record = outputData(nextRecordId) // existst as ist is firstkey
        outputData -= nextRecordId
        writer.println(record.csvOriginal)
        recordsWritten += 1
        nextRecordId += 1
        if (0 == (nextRecordId % 10000)) log.debug(s"processed: ${nextRecordId}")
    }
      if (recordsWritten > 0) {
        reader ! RecordsWritten(recordsWritten)
      }
    }
  }
}

object Writer {
  def props() = Props(new Writer())
}
